sql/src/main/java/io/crate/execution/engine/indexing/ShardingUpsertExecutor.java

```diff
@@ -339,19 +346,41 @@ private void maybeSetInterrupt(@Nullable Exception failure) {
     private static class HeapThresholdChecker implements Predicate<ShardedRequests<?, ?>> {
 
         private final Runtime rt;
-        private final AtomicLong threshold;
+        private final long threshold;
+        private final List<Symbol> columnSymbols;
+        private final GroupRowsByShard<ShardUpsertRequest, ShardUpsertRequest.Item> grouper;
+        private final AtomicLong memInFlight;
 
-        HeapThresholdChecker(int freeHeapPercent) {
+        HeapThresholdChecker(int freeHeapPercent, List<Symbol> columnSymbols, GroupRowsByShard<ShardUpsertRequest, ShardUpsertRequest.Item> grouper) {
```

Comment:
Maybe we split this up a bit so that the `GroupRowsByShard` doesn't need to expose the `getCurrentRowExpressions`, instead we expose a new  `estimatedSize` property on the `shardedRequests` and move the calculation for it elsewhere:

- We create a `Function<Row, Long>` implementation that encapsulates the `SizeEstimator` usage (we don't want to re-create the estimators per row, but only once
- We utilize this new Function in `GroupRowsByShard`, this receives the row in each `accept` call, and can add the estimated size of the rows to the `ShardedRequests` via the `add` call it does when it adds the resolved parts from the row.
- We expose the accumulated estimated size in  `ShardedRequests` (which will also be reset for each new partition, since we're creating new `ShardedRequests` instances, which is what we want.